In [36]:
import numpy as np
np.random.seed(42)
from sklearn import datasets

### Dataset preparation

In [37]:
iris = datasets.load_iris()

x = iris.data[:, [0, 1]]
x = iris.data[:, [0, 2]]
x = iris.data[:, [0, 3]]
x = iris.data[:, [1, 2]]
x = iris.data[:, [1, 3]]

y = iris.target

class_names = iris.target_names
discription = iris.DESCR

### Dataset Split

In [38]:
num_samples = x.shape[0]
num_features = x.shape[1]

print(f"num_samples: {num_samples}")
print(f"num_features: {num_features}")

num_samples: 150
num_features: 2


In [39]:
test_size = num_samples // 3
random_idxs = np.random.permutation(num_samples)

x_train = x[random_idxs[:-test_size]]
y_train = y[random_idxs[:-test_size]]

x_test = x[random_idxs[-test_size:]]
y_test = y[random_idxs[-test_size:]]

In [40]:
print(f"x_train shape:\n{x_train.shape}")
print(f"y_train shape:\n{y_train.shape}")

print(f"x_test shape:\n{x_test.shape}")
print(f"y_test shape:\n{y_test.shape}")

x_train shape:
(100, 2)
y_train shape:
(100,)
x_test shape:
(50, 2)
y_test shape:
(50,)


### KNN Model

In [41]:
from sklearn.neighbors import KNeighborsClassifier

clf = KNeighborsClassifier(n_neighbors=3)
clf.fit(x_train, y_train)

accuracy = clf.score(x_test, y_test)
print(f"accuracy: {accuracy*100.0:.4}%")

y_pred = clf.predict(x_test)
print(f"y_pred:\n{y_pred}")

accuracy: 98.0%
y_pred:
[0 0 0 1 2 0 2 2 0 1 1 2 1 2 0 2 1 2 1 1 1 0 1 1 0 1 2 2 0 1 2 1 0 2 0 1 2
 2 1 2 1 1 2 2 0 1 2 0 1 2]


### Try different hyperparameters

In [42]:
n_neighbors = [n_neighbor for n_neighbor in range(1, 11)] # [1, 2, ..., 10]
weight_modes = ['uniform', 'distance']

best_acc = 0.0

# 10 x 2 = 20 Models
for n_neighbor in n_neighbors:
    for weight_mode in weight_modes:
        clf = KNeighborsClassifier(
            n_neighbors=n_neighbor,
            weights=weight_mode
        )
        clf.fit(x_train, y_train)
        accuracy = clf.score(x_test, y_test)
        if accuracy > best_acc:
            best_acc = accuracy
            print(
                f"Neighbors: {n_neighbor}  "
                f"Weight: {weight_mode}  "
                f"Score: {accuracy:.4}"
            )

Neighbors: 1  Weight: uniform  Score: 0.94
Neighbors: 2  Weight: uniform  Score: 0.98


### Performances

- 0, 1: Neighbors: 8  Weight: uniform  Score: 0.74
- 0, 2: Neighbors: 5  Weight: uniform  Score: 0.96
- 0, 3: Neighbors: 4  Weight: uniform  Score: 0.96
- 1, 2: Neighbors: 10  Weight: uniform  Score: 0.96
- 1, 3: Neighbors: 2  Weight: uniform  Score: 0.98